sympy/tensor/array/array_comprehension.py

```diff
@@ -329,3 +332,68 @@ def tomatrix(self):
             raise ValueError('Dimensions must be of size of 2')
 
         return Matrix(self._expand_array())
+
+
+def isLambda(v):
+    LAMBDA = lambda: 0
+    return isinstance(v, type(LAMBDA)) and v.__name__ == LAMBDA.__name__
+
+class ArrayComprehensionMap(ArrayComprehension):
+    '''
+    A subclass of ArrayComprehension dedicated to map external function lambda.
+
+    Notes
+    =====
+
+    Only the lambda function is considered.
+    At most one argument in lambda function is accepted in order to avoid ambiguity.
+    Function subs() is not supported, so all the bound should be numeric.
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array import ArrayComprehensionMap
+    >>> from sympy import symbols
+    >>> i, j, k = symbols('i j k')
+    >>> a = ArrayComprehensionMap(lambda: 1, (i, 1, 4))
+    >>> a.doit()
+    [1, 1, 1, 1]
+    >>> b = ArrayComprehensionMap(lambda a: a+1, (j, 1, 4))
+    >>> b.doit()
+    [2, 3, 4, 5]
+
+    '''
+    def __new__(cls, function, *symbols, **assumptions):
+        if any(len(l) != 3 or None for l in symbols):
+            raise ValueError('ArrayComprehension requires values lower and upper bound'
+                              ' for the expression')
+
+        for _, inf, sup in symbols:
+            if Basic(inf, sup).atoms(Symbol):
+                raise ValueError('Only numeric bound is accepted')
+
+        arglist = [Expr('LAMBDA')]
+        arglist.extend(cls._check_limits_validity(function, symbols))
+        obj = cls._new(arglist, **assumptions)
+        obj._lambda = function
+        return obj
+
+    def _expand_array(self):
+        list_gen = self._lambda
+        for var, inf, sup in reversed(self._limits):
+            list_expr = list_gen
+            list_gen = []
+            for val in range(inf, sup+1):
+                if not isinstance(list_expr, Iterable):
+                    if isLambda(list_expr):
+                        if list_expr.__code__.co_argcount == 0:
+                            list_gen.append(list_expr())
+                        elif list_expr.__code__.co_argcount == 1:
+                            list_gen.append(list_expr(val))
+                        else:
+                            raise ValueError("One argument at most is accepted")
+                    else:
+                        list_gen.append(list_expr.subs(var, val))
+                else:
+                    list_gen.append(self._array_subs(list_expr, var, val))
+        return list_gen
```

Comment:
You need to overwrite `def func(self, *args)` because `function` won't be in `.args`. Try something like this:
```python
@property
def func(self):
    class _(ArrayComprehensionMap):
        def __new__(cls, *args, **kwargs):
            return ArrayComprehensionMap(self._lambda, *args, **kwargs)
    return _
```

Hope I got the code right... `func` needs to return a class object, otherwise SymPy will get angry.